### PR TITLE
[mso_schema_site_external_epg] Adding new mso_schema_site_external_epg module for non-cloud sites. Fixes #119

### LINKED ISSUE #119: [mso_schema_site_external_epg] Add capability to set L3Out mapping at Site Level
Currently External EPG to L3Out can be mapped at template level but not at Site level.
This ask is to add a new module called mso_schema_site_external_epg allowing to set that mapping.

### PR DESCRIPTION
**Pull request by anvitha-jain**
Resolves issue #119 by adding new module (mso_schema_site_external_epg) to add external EPGs to non-cloud sites.

### CODE REVIEW

File: plugins/modules/mso_schema_template.py [hunk 1]
```diff
@@ -145,10 +147,17 @@ def main():
...
         if template:
             if template in templates:
+                # mso.stdout = str(mso.get_obj('template')
                 template_idx = templates.index(template)
                 mso.existing = schema_obj.get('templates')[template_idx]
+                mso.stdout = str(mso.existing['tenantId'])
+                path = 'tenants/{0}'.format(mso.existing['tenantId'])
+                mso.stdout += str(path)
+                mso.stdout += str('-----------')
+                mso.stdout += str(mso.query_objs(path)['name'])
```
Comment (shrsr) [inline plugins/modules/mso_schema_template.py:160]: Please remove all the print statements (mso.stdout)

File: plugins/modules/mso_schema_template.py [hunk 2]
```diff
@@ -145,10 +147,17 @@ def main():
 
         # Get template
         templates = [t.get('name') for t in schema_obj.get('templates')]
+        # mso.stdout = str(mso.get_obj('templates', name=template))
```
Comment (shrsr) [inline plugins/modules/mso_schema_template.py:150]: Remove the comment

File: plugins/modules/mso_schema_template.py [hunk 3]
```diff
@@ -135,6 +135,8 @@ def main():
 
     mso = MSOModule(module)
 
+    # mso.stdout = str(mso.get_obj('templates', name=template))
```
Comment (shrsr) [inline plugins/modules/mso_schema_template.py:138]: Remove the comment

File: plugins/modules/mso_schema_template.py [hunk 4]
```diff
@@ -135,6 +135,8 @@ def main():
 
     mso = MSOModule(module)
```
Comment (shrsr) [inline plugins/modules/mso_schema_template.py:136]: Can you revert this module to the original?

File: plugins/modules/mso_schema_site_external_epg_selector.py [hunk 5]
```diff
@@ -185,6 +185,7 @@ def main():
     if template not in templates:
         mso.fail_json(msg="Provided template '{template}' does not exist. Existing templates: {templates}".format(template=template,
                                                                                                                   templates=', '.join(templates)))
+    mso.stdout = str(mso.get_obj('templates', name=template))
```
Comment (shrsr) [inline plugins/modules/mso_schema_site_external_epg_selector.py:188]: not required

**Review by shrsr — CHANGES_REQUESTED** [review]
Please see comments

File: plugins/modules/mso_schema_template.py (hunk 2 above)
Comment (anvitha-jain) [reply plugins/modules/mso_schema_template.py:150]: done

File: plugins/modules/mso_schema_site_external_epg_selector.py (hunk 5 above)
Comment (anvitha-jain) [reply plugins/modules/mso_schema_site_external_epg_selector.py:188]: changed

File: plugins/modules/mso_schema_template.py (hunk 4 above)
Comment (anvitha-jain) [reply plugins/modules/mso_schema_template.py]: done

**Comment by anvitha-jain** [comment]
Cleaned code. Won't work due to API.

File: plugins/modules/mso_schema_site_external_epg.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        schema=dict(type='str', required=True),
+        template=dict(type='str', required=True),
+        site=dict(type='str', required=True),
+        vrf=dict(type='str'),
+        l3out=dict(type='str'),
+        external_epg=dict(type='str'),
```
Comment (lhercot) [inline plugins/modules/mso_schema_site_external_epg.py:81]: We generally have 'name' as alias of the main parameter.

File: plugins/modules/mso_schema_site_external_epg.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+    if ext_epg_ref in external_epgs:
+        external_epg_idx = external_epgs.index(ext_epg_ref)
+        # Get External EPG
+        mso.existing = schema_obj['sites'][site_idx]['externalEpgs'][external_epg_idx]
+        op_path = '/sites/{0}/externalEpgs/{1}'.format(site_template, external_epg)
+
+    ops = []
+
+    mso.stdout = str(external_epg)
```
Comment (lhercot) [inline plugins/modules/mso_schema_site_external_epg.py:155]: Remove all mso.stdout statements.

File: plugins/modules/mso_schema_site_l3out.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        schema=dict(type='str', required=True),
+        site=dict(type='str', required=True),
+        template=dict(type='str', required=True),
+        vrf=dict(type='dict', options=mso_reference_spec()),
+        description=dict(type='str'),
+        display_name=dict(type='str'),
+        l3out=dict(type='str'),  # This parameter is not required for querying all objects
```
Comment (lhercot) [inline plugins/modules/mso_schema_site_l3out.py:154]: This has generally name as an alias.

**Review by lhercot — CHANGES_REQUESTED** [review]
Seems you also created an error on MSO 2.2.4 check the integration tasks for more details but I see the following error on MSO 2.2.4:
``` json
{
  "code": 500,
  "message": "JsResultException(errors:List((/_id/$oid(0)(0),List(ValidationError(List('l3outRef' is undefined on object: {\"externalEpgRef\":{\"schemaId\":\"60a3a842110000ae2e30ec17\",\"templateName\":\"Template1\",\"l3outName\":\"L3out1\"},\"vrfRef\":{\"schemaId\":\"60a3a842110000ae2e30ec17\",\"templateName\":\"Template1\",\"vrfName\":\"VRF1\"}}),WrappedArray())))))"
}
```

File: plugins/modules/mso_schema_site_external_epg.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: mso_schema_site_external_epg
+short_description: Manage External EPG in schema of cloud sites
+description:
+- Manage External EPG in schema of cloud sites on Cisco ACI Multi-Site.
```
Comment (lhercot) [inline plugins/modules/mso_schema_site_external_epg.py:19]: Is it for Cloud Sites only? I would remove the "cloud" from this line.

File: plugins/modules/mso_schema_site_external_epg.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+- This module can only be used on versions of MSO that are 3.1.1h or greater.
```
Comment (lhercot) [inline plugins/modules/mso_schema_site_external_epg.py:20]: Isn't this only for 3.3 or greater?

File: plugins/modules/mso_schema_site_external_epg.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    - The L3Out associated with the external epg.
+    type: str
+  external_epg:
+    description:
+    - The name of the External EPG to be managed.
+    type: str
+    aliases: [ name ]
+  site:
+    description:
+    - The name of the cloud site.
```
Comment (lhercot) [inline plugins/modules/mso_schema_site_external_epg.py:45]: I would remove cloud from this description

File: plugins/modules/mso_schema_site_l3out.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: mso_schema_site_l3out
+short_description: Manage site-local layer3 Out (L3Outs) in schema template
+description:
+- Manage site-local L3Outs in schema template on Cisco ACI Multi-Site.
+- This module can only be used on versions of MSO that are 3.0 or greater.
+- NOTE- Usage of this module for version lesser than 3.0 might break the MSO.
```
Comment (lhercot) [inline plugins/modules/mso_schema_site_l3out.py:21]: missing space after NOTE

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**